File: packages/gbnf-experiment/data/20260909T151143Z_170b737d/ported_implementation/src/rules_builder/rules_builder.ts
import { GrammarParseError } from "../utils/errors/index.ts";
import { at } from "../utils/python_compat.ts";
import { is_word_char } from "./is_word_char.ts";
import { parse_char } from "./parse_char.ts";
import { parse_name } from "./parse_name.ts";
import { parse_space } from "./parse_space.ts";
import {
  InternalRuleDefAlt,
  InternalRuleDefChar,
  InternalRuleDefCharAlt,
  InternalRuleDefCharNot,
  InternalRuleDefCharRngUpper,
  InternalRuleDefEnd,
  InternalRuleDefReference,
  type InternalRuleDef,
} from "./rules_builder_types.ts";
import { SymbolIds } from "./symbol_ids.ts";

type InternalRuleDefConstructor =
  | typeof InternalRuleDefChar
  | typeof InternalRuleDefCharNot
  | typeof InternalRuleDefCharRngUpper
  | typeof InternalRuleDefAlt
  | typeof InternalRuleDefEnd
  | typeof InternalRuleDefCharAlt;

export const get_out_elements = (
  type_of_rule: InternalRuleDefConstructor,
  startchar_value: number,
): InternalRuleDef => {
  if (type_of_rule === InternalRuleDefChar) {
    return new InternalRuleDefChar([startchar_value]);
  }
  if (type_of_rule === InternalRuleDefCharNot) {
    return new InternalRuleDefCharNot([startchar_value]);
  }
  if (type_of_rule === InternalRuleDefCharRngUpper) {
    return new InternalRuleDefCharRngUpper(startchar_value);
  }
  if (type_of_rule === InternalRuleDefAlt) {
    return new InternalRuleDefAlt();
  }
  if (type_of_rule === InternalRuleDefEnd) {
    return new InternalRuleDefEnd();
  }
  if (type_of_rule === InternalRuleDefCharAlt) {
    return new InternalRuleDefCharAlt(startchar_value);
  }

  throw new Error(`Invalid type: ${type_of_rule}`);
};

const is_python_space = (char: string): boolean => /\s/.test(char);

export class RulesBuilder {
  pos: number;
  symbol_ids: SymbolIds;
  rules: InternalRuleDef[][];
  src: string;
  start: number;
  time_limit: number;

  constructor(src: string, limit: number = 1000) {
    this.pos = 0;
    this.symbol_ids = new SymbolIds();
    this.rules = [];
    this.src = src;
    this.start = performance.now() / 1000;
    this.time_limit = limit;
    this.parse(src);
  }

  parse(src: string): void {
    this.pos = parse_space(src, 0, true);
    while (this.pos < src.length) {
      this.parse_rule(src);
    }

    // Validate the state to ensure that all rules are defined
    for (const rule of this.rules) {
      for (const elem of rule) {
        if (elem instanceof InternalRuleDefReference) {
          const rule_exists =
            elem.value < this.rules.length && this.rules[elem.value].length > 0;
          if (!rule_exists) {
            const missing_rule_name = this.symbol_ids.reverse_get(elem.value);
            let missing_rule_pos = this.symbol_ids.get_pos(missing_rule_name);

            // Skip over the ::= and any whitespace
            while (
              missing_rule_pos < src.length &&
              (src[missing_rule_pos] === ":" ||
                src[missing_rule_pos] === "=" ||
                is_python_space(src[missing_rule_pos]))
            ) {
              missing_rule_pos += 1;
            }

            throw new GrammarParseError(
              src,
              missing_rule_pos,
              `Undefined rule identifier "${missing_rule_name}"`,
            );
          }
        }
      }
    }
  }

  parse_rule(src: string): void {
    const name = parse_name(src, this.pos);
    this.pos = parse_space(src, this.pos + name.length, false);
    const rule_id = this.get_symbol_id(name, name.length);

    this.pos = parse_space(src, this.pos, true);
    if (
      !(
        this.pos + 2 < src.length && // Ensure the position + 2 is within bounds
        src[this.pos] === ":" &&
        src[this.pos + 1] === ":" &&
        src[this.pos + 2] === "="
      )
    ) {
      throw new GrammarParseError(src, this.pos, `Expecting ::= at ${this.pos}`);
    }
    this.pos += 3;
    this.pos = parse_space(src, this.pos, true);

    this.parse_alternates(name, rule_id);

    // Check if this.pos is within the bounds of src before checking for a carriage return
    if (this.pos < src.length && src[this.pos] === "\r") {
      this.pos += at(src, this.pos + 1) === "\n" ? 2 : 1;
    } else if (this.pos < src.length && src[this.pos] === "\n") {
      this.pos += 1;
    } else if (this.pos < src.length && src[this.pos]) {
      throw new GrammarParseError(src, this.pos, `Expecting newline or end at ${this.pos}`);
    }
    this.pos = parse_space(src, this.pos, true);
  }

  get_symbol_id(src: string, length: number): number {
    const next_id = this.symbol_ids.length;
    const key = src.slice(0, length);
    if (!this.symbol_ids.has(key)) {
      this.symbol_ids.set(key, next_id, this.pos);
    }
    return this.symbol_ids.getItem(key);
  }

  generate_symbol_id(base_name: string): number {
    const next_id = this.symbol_ids.length;
    this.symbol_ids.set(`${base_name}_${next_id}`, next_id, this.pos);
    return next_id;
  }

  add_rule(rule_id: number, rule: InternalRuleDef[]): void {
    while (this.rules.length <= rule_id) {
      this.rules.push([]);
    }
    this.rules[rule_id] = rule;
  }

  check_duration(): void {
    if (performance.now() / 1000 - this.start > this.time_limit) {
      throw new GrammarParseError(this.src, this.pos, `Duration of ${this.time_limit} exceeded`);
    }
  }

  parse_sequence(rule_name: string, out_elements: InternalRuleDef[], depth: number = 0): void {
    const is_nested = depth !== 0;
    const src = this.src;
    let last_sym_start = out_elements.length;

    while (this.pos < src.length) {
      if (src[this.pos] === '"') {
        this.pos += 1;
        last_sym_start = out_elements.length;
        while (at(src, this.pos) !== '"') {
          this.check_duration();
          const [value, inc_pos] = parse_char(src, this.pos);
          out_elements.push(new InternalRuleDefChar([value]));
          this.pos += inc_pos;
        }
        this.pos = parse_space(src, this.pos + 1, is_nested);
      } else if (src[this.pos] === "[") {
        this.pos += 1;
        let start_type: InternalRuleDefConstructor = InternalRuleDefChar;
        if (at(src, this.pos) === "^") {
          this.pos += 1;
          start_type = InternalRuleDefCharNot;
        }
        last_sym_start = out_elements.length;
        while (at(src, this.pos) !== "]") {
          this.check_duration();
          const type_: InternalRuleDefConstructor =
            last_sym_start < out_elements.length ? InternalRuleDefCharAlt : start_type;
          const [startchar_value, inc_pos] = parse_char(src, this.pos);
          this.pos += inc_pos;
          out_elements.push(get_out_elements(type_, startchar_value));

          if (at(src, this.pos) === "-" && at(src, this.pos + 1) !== "]") {
            this.pos += 1;
            const [endchar_value, end_inc_pos] = parse_char(src, this.pos);
            out_elements.push(new InternalRuleDefCharRngUpper(endchar_value));
            this.pos += end_inc_pos;
          }
        }
        this.pos = parse_space(src, this.pos + 1, is_nested);
      } else if (is_word_char(src[this.pos])) {
        const name = parse_name(src, this.pos);
        const ref_rule_id = this.get_symbol_id(name, name.length);
        this.pos += name.length;
        this.pos = parse_space(src, this.pos, is_nested);

        last_sym_start = out_elements.length;
        out_elements.push(new InternalRuleDefReference(ref_rule_id));
      } else if (src[this.pos] === "(") {
        this.pos = parse_space(src, this.pos + 1, true);
        const sub_rule_id = this.generate_symbol_id(rule_name);
        this.parse_alternates(rule_name, sub_rule_id, depth + 1);
        last_sym_start = out_elements.length;
        out_elements.push(new InternalRuleDefReference(sub_rule_id));
        if (at(src, this.pos) !== ")") {
          throw new GrammarParseError(src, this.pos, `Expecting ')' at ${this.pos}`);
        }
        this.pos = parse_space(src, this.pos + 1, is_nested);
      } else if ("*+?".includes(src[this.pos])) {
        if (last_sym_start === out_elements.length) {
          throw new GrammarParseError(
            src,
            this.pos,
            `Expecting preceding item to */+/? at ${this.pos}`,
          );
        }
        const sub_rule_id = this.generate_symbol_id(rule_name);
        const sub_rule = out_elements.slice(last_sym_start);
        if ("*+".includes(src[this.pos])) {
          sub_rule.push(new InternalRuleDefReference(sub_rule_id));
        }
        sub_rule.push(new InternalRuleDefAlt());
        if (src[this.pos] === "+") {
          sub_rule.push(...out_elements.slice(last_sym_start));
        }
        sub_rule.push(new InternalRuleDefEnd());
        this.add_rule(sub_rule_id, sub_rule);
        out_elements.splice(
          last_sym_start,
          out_elements.length - last_sym_start,
          new InternalRuleDefReference(sub_rule_id),
        );
        this.pos = parse_space(src, this.pos + 1, is_nested);
      } else {
        break;
      }
    }
  }

  parse_alternates(rule_name: string, rule_id: number, depth: number = 0): void {
    const src = this.src;
    const rule: InternalRuleDef[] = [];
    this.parse_sequence(rule_name, rule, depth);
    // Ensure that this.pos is within bounds before checking src[this.pos]
    while (this.pos < src.length && src[this.pos] === "|") {
      this.check_duration();
      rule.push(new InternalRuleDefAlt());
      this.pos = parse_space(src, this.pos + 1, true);
      this.parse_sequence(rule_name, rule, depth);
    }
    rule.push(new InternalRuleDefEnd());
    this.add_rule(rule_id, rule);
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T151143Z_170b737d/20260910T085124Z_7f672549/ported_implementation/gbnf/rules_builder/rules_builder.py
import time
from typing import List

from ..utils.errors import GrammarParseError
from .is_word_char import is_word_char
from .parse_char import parse_char
from .parse_name import parse_name
from .parse_space import parse_space
from .rules_builder_types import (
    InternalRuleDef,
    InternalRuleDefAlt,
    InternalRuleDefChar,
    InternalRuleDefCharAlt,
    InternalRuleDefCharNot,
    InternalRuleDefCharRngUpper,
    InternalRuleDefEnd,
    InternalRuleDefReference,
)
from .symbol_ids import SymbolIds


def get_out_elements(type_of_rule, startchar_value: int) -> InternalRuleDef:
    if type_of_rule is InternalRuleDefChar:
        return InternalRuleDefChar([startchar_value])
    if type_of_rule is InternalRuleDefCharNot:
        return InternalRuleDefCharNot([startchar_value])
    if type_of_rule is InternalRuleDefCharRngUpper:
        return InternalRuleDefCharRngUpper(startchar_value)
    if type_of_rule is InternalRuleDefAlt:
        return InternalRuleDefAlt()
    if type_of_rule is InternalRuleDefEnd:
        return InternalRuleDefEnd()
    if type_of_rule is InternalRuleDefCharAlt:
        return InternalRuleDefCharAlt(startchar_value)

    raise ValueError(f"Invalid type: {type_of_rule}")


class RulesBuilder:
    def __init__(self, src: str, limit: int = 1000):
        self.pos = 0
        self.symbol_ids = SymbolIds()
        self.rules: List[List[InternalRuleDef]] = []
        self.src = src
        self.start = time.time()
        self.time_limit = limit
        self.parse(src)

    def parse(self, src: str) -> None:
        self.pos = parse_space(src, 0, True)
        while self.pos < len(src):
            self.parse_rule(src)

        # Validate the state to ensure that all rules are defined
        for rule in self.rules:
            for elem in rule:
                if isinstance(elem, InternalRuleDefReference):
                    rule_exists = (
                        elem.value < len(self.rules)
                        and len(self.rules[elem.value]) > 0
                    )
                    if not rule_exists:
                        missing_rule_name = self.symbol_ids.reverse_get(elem.value)
                        missing_rule_pos = self.symbol_ids.get_pos(missing_rule_name)

                        # Skip over the ::= and any whitespace
                        while missing_rule_pos < len(src) and (
                            src[missing_rule_pos] == ":"
                            or src[missing_rule_pos] == "="
                            or src[missing_rule_pos].isspace()
                        ):
                            missing_rule_pos += 1

                        raise GrammarParseError(
                            src,
                            missing_rule_pos,
                            f'Undefined rule identifier "{missing_rule_name}"',
                        )

    def parse_rule(self, src: str) -> None:
        name = parse_name(src, self.pos)
        self.pos = parse_space(src, self.pos + len(name), False)
        rule_id = self.get_symbol_id(name, len(name))

        self.pos = parse_space(src, self.pos, True)
        if not (
            self.pos + 2 < len(src)  # Ensure the position + 2 is within bounds
            and src[self.pos] == ":"
            and src[self.pos + 1] == ":"
            and src[self.pos + 2] == "="
        ):
            raise GrammarParseError(src, self.pos, f"Expecting ::= at {self.pos}")
        self.pos += 3
        self.pos = parse_space(src, self.pos, True)

        self.parse_alternates(name, rule_id)

        # Check if self.pos is within the bounds of src before checking for a carriage return
        if self.pos < len(src) and src[self.pos] == "\r":
            self.pos += 2 if src[self.pos + 1] == "\n" else 1
        elif self.pos < len(src) and src[self.pos] == "\n":
            self.pos += 1
        elif self.pos < len(src) and src[self.pos]:
            raise GrammarParseError(
                src, self.pos, f"Expecting newline or end at {self.pos}"
            )
        self.pos = parse_space(src, self.pos, True)

    def get_symbol_id(self, src: str, length: int) -> int:
        next_id = len(self.symbol_ids)
        key = src[0:length]
        if key not in self.symbol_ids:
            self.symbol_ids.set(key, next_id, self.pos)
        return self.symbol_ids[key]

    def generate_symbol_id(self, base_name: str) -> int:
        next_id = len(self.symbol_ids)
        self.symbol_ids.set(f"{base_name}_{next_id}", next_id, self.pos)
        return next_id

    def add_rule(self, rule_id: int, rule: List[InternalRuleDef]) -> None:
        while len(self.rules) <= rule_id:
            self.rules.append([])
        self.rules[rule_id] = rule

    def check_duration(self) -> None:
        if time.time() - self.start > self.time_limit:
            raise GrammarParseError(
                self.src, self.pos, f"Duration of {self.time_limit} exceeded"
            )

    def parse_sequence(
        self, rule_name: str, out_elements: List[InternalRuleDef], depth: int = 0
    ) -> None:
        is_nested = depth != 0
        src = self.src
        last_sym_start = len(out_elements)

        while self.pos < len(src):
            if src[self.pos] == '"':
                self.pos += 1
                last_sym_start = len(out_elements)
                while src[self.pos] != '"':
                    self.check_duration()
                    value, inc_pos = parse_char(src, self.pos)
                    out_elements.append(InternalRuleDefChar([value]))
                    self.pos += inc_pos
                self.pos = parse_space(src, self.pos + 1, is_nested)
            elif src[self.pos] == "[":
                self.pos += 1
                start_type = InternalRuleDefChar
                if src[self.pos] == "^":
                    self.pos += 1
                    start_type = InternalRuleDefCharNot
                last_sym_start = len(out_elements)
                while src[self.pos] != "]":
                    self.check_duration()
                    type_ = (
                        InternalRuleDefCharAlt
                        if last_sym_start < len(out_elements)
                        else start_type
                    )
                    startchar_value, inc_pos = parse_char(src, self.pos)
                    self.pos += inc_pos
                    out_elements.append(get_out_elements(type_, startchar_value))

                    if src[self.pos] == "-" and src[self.pos + 1] != "]":
                        self.pos += 1
                        endchar_value, end_inc_pos = parse_char(src, self.pos)
                        out_elements.append(
                            InternalRuleDefCharRngUpper(endchar_value)
                        )
                        self.pos += end_inc_pos
                self.pos = parse_space(src, self.pos + 1, is_nested)
            elif is_word_char(src[self.pos]):
                name = parse_name(src, self.pos)
                ref_rule_id = self.get_symbol_id(name, len(name))
                self.pos += len(name)
                self.pos = parse_space(src, self.pos, is_nested)

                last_sym_start = len(out_elements)
                out_elements.append(InternalRuleDefReference(ref_rule_id))
            elif src[self.pos] == "(":
                self.pos = parse_space(src, self.pos + 1, True)
                sub_rule_id = self.generate_symbol_id(rule_name)
                self.parse_alternates(rule_name, sub_rule_id, depth + 1)
                last_sym_start = len(out_elements)
                out_elements.append(InternalRuleDefReference(sub_rule_id))
                if src[self.pos] != ")":
                    raise GrammarParseError(
                        src, self.pos, f"Expecting ')' at {self.pos}"
                    )
                self.pos = parse_space(src, self.pos + 1, is_nested)
            elif src[self.pos] in "*+?":
                if last_sym_start == len(out_elements):
                    raise GrammarParseError(
                        src,
                        self.pos,
                        f"Expecting preceding item to */+/? at {self.pos}",
                    )
                sub_rule_id = self.generate_symbol_id(rule_name)
                sub_rule = out_elements[last_sym_start:]
                if src[self.pos] in "*+":
                    sub_rule.append(InternalRuleDefReference(sub_rule_id))
                sub_rule.append(InternalRuleDefAlt())
                if src[self.pos] == "+":
                    sub_rule.extend(out_elements[last_sym_start:])
                sub_rule.append(InternalRuleDefEnd())
                self.add_rule(sub_rule_id, sub_rule)
                out_elements[last_sym_start:] = [
                    InternalRuleDefReference(sub_rule_id)
                ]
                self.pos = parse_space(src, self.pos + 1, is_nested)
            else:
                break

    def parse_alternates(self, rule_name: str, rule_id: int, depth: int = 0) -> None:
        src = self.src
        rule: List[InternalRuleDef] = []
        self.parse_sequence(rule_name, rule, depth)
        # Ensure that self.pos is within bounds before checking src[self.pos]
        while self.pos < len(src) and src[self.pos] == "|":
            self.check_duration()
            rule.append(InternalRuleDefAlt())
            self.pos = parse_space(src, self.pos + 1, True)
            self.parse_sequence(rule_name, rule, depth)
        rule.append(InternalRuleDefEnd())
        self.add_rule(rule_id, rule)
